### DISCUSSION TITLE
Model execution time for a 600x800 color image takes 7.5 seconds - Any ways to reduce ??

### DISCUSSION THREAD

**Discussion by gganes3** [mindee/doctr#814] [Q&A]
Using db resnet-50 and crnnvgg16_bn models for prediction, is there any way to reduce execution time?

**Comment by fg-mindee** [comment]
Hi @gganes3 :wave: 

This is quite an interesting topic :+1: 
There are many things at hand, I'll try to give the best overview I can:
- **input size** : the input size has a direct impact on latency. It's recommended to start with the size the model was trained on, but if needed, you can try to reduce the input size (that might have some consequences on performances, depending on how far you changed it to)
- **hardware**: from the latency you measured, I guess you are running this on CPU. You could switch to GPU for a better latency!
- **backend** : which backend are you running? Pytorch or TensorFlow?
- **architecture choice**: docTR provides several architectures, and depending on whether you run PyTorch or TensorFlow, the layers of the model will not have the same latency (the under-the-hood low-level implementations of layers is specific to the framework/backend). But you could check lighter architectures to get a better latency
- **runtime**: so this would require doing a few these manually (cannot be used as is) but if using a DL framework for training or development doesn't mean that one cannot use something else for inference. Pytorch models usually convert nicely to [ONNX](https://onnxruntime.ai/) for instance, which is usually better for CPU in terms of latency. For TensorFlow there are a few options, but not supported officially as of today: you can check [OpenVINO](https://docs.openvino.ai/latest/index.html)


Now that being said, could you share a minimal snippet showcasing how you performed inference please?
The default size that we use is (1024, 1024), and as you can see on our latency benchmarks (https://github.com/mindee/doctr/blob/main/references/detection/latency.csv, https://github.com/mindee/doctr/blob/main/references/recognition/latency.csv), the measured latency is way lower than this. This could be because you have an extremely dense document, and thus that the recognition model needs to run on many batches.

I hope this helps :)

**Reply by gganes3** [reply]
I am using pytorch
        det_model = db_resnet50(pretrained=False)
        reco_model = crnn_vgg16_bn(pretrained=False)

        det_params = torch.load(r"C:\Projects\db_resnet50-ac60cadc.pt", map_location="cpu")
        reco_params = torch.load(r"C:\Projects\crnn_vgg16_bn-9762b0b0.pt", map_location="cpu")
        # det_model.load_state_dict()
        det_model.load_state_dict(det_params)
        reco_model.load_state_dict(reco_params)

        # Ask the preprocessor of each task to resize and normalize similarly to your training
        # cf. https://github.com/mindee/doctr/blob/main/references/detection/train_pytorch.py#L94 & https://github.com/mindee/doctr/blob/main/references/detection/train_pytorch.py#L109
        det_predictor = DetectionPredictor(
            PreProcessor((1024, 1024), batch_size=1, mean=(0.798, 0.785, 0.772), std=(0.264, 0.2749, 0.287)), det_model)
        # cf. https://github.com/mindee/doctr/blob/main/references/recognition/train_pytorch.py#L97 & https://github.com/mindee/doctr/blob/main/references/recognition/train_pytorch.py#L111
        reco_predictor = RecognitionPredictor(
            PreProcessor((32, 128), preserve_aspect_ratio=True, batch_size=32, mean=(0.694, 0.695, 0.693),
                         std=(0.299, 0.296, 0.301)), reco_model)
        print("Time Consumed for model loading")
        print("% s seconds" % (time.time() - load_mdl_strt))
        start_mdl = time.time()
        model = OCRPredictor(det_predictor, reco_predictor)
        image_rgb = [image]
        result = model(image_rgb)
        
        This is how I am doing inference.

**Reply by fg-mindee** [reply]
First, I'd move the `model = OCRPredictor(det_predictor, reco_predictor)` in the model loading section!

And about the other aspects, what hardware are you running this on? How dense are those documents?

**Reply by gganes3** [reply]
Sorry, I didn't understand what you meant by moving the model = OCRPredictor(det_predictor,reco_predictor). hardware is a CPU with 16GB RAM, 2GB GPU Windows OS. Actually, I am running it on images ranging with sizes of 600x800x3 to 4096x4096x3.

**Reply by fg-mindee** [reply]
I meant moving that line of code up haha
It's instantiating the end-to-end predictor so that shouldn't be done every time you perform inference

Alright, so the GPU is not very beefy I guess (considering the RAM), but for the CPU, how powerful is it? (is this an Intel i7? which gen?)

For the density of the document, I don't mean the size of the image, I mean the number of words in the page. The higher this number, the more inputs will be fed to the recognition model ;)

**Reply by gganes3** [reply]
Its a combination of Alphabets and numerals. Maximum of 80 alphabets and numerals together.

**Reply by fg-mindee** [reply]
So you mean that there are only about 80 words on your image?
Would you mind sharing your input image? Because in our experiments, the inference is much faster :thinking:

**Reply by gganes3** [reply]
yes. However, I can't share image because it is customer confidential. I was playing with reducing the image to be in a rectangular size (256x1024) and it seems to be working okay. Similarly the detection model also, I changed to 256x1024. Is this advisable for a longer run?

**Reply by fg-mindee** [reply]
If you are retraining the model, and your use case only includes rectangular images, yes sure!

But considering we've been able to run the end-to-end pipeline on 1024x1024 for the detection part at a much lower latency, I'm concerned we might be missing something here :thinking:

**Reply by gganes3** [reply]
Yeah, same thoughts here. Not able to understand why. I am trying to move to the GPU however I am encountering an issue where it says Torch not compiled with CUDA enabled
```
  File "C:/Projects/AutoPixelSizeCalculation/AutoPixelSizeCalculation_ocrpredictor_test.py", line 48, in calculatepixelsize
    model = OCRPredictor(det_predictor, reco_predictor).cuda()
  File "C:\Users\gganesan166268\AppData\Local\Programs\Python\Python37\lib\site-packages\torch\nn\modules\module.py", line 680, in cuda
    return self._apply(lambda t: t.cuda(device))
  File "C:\Users\gganesan166268\AppData\Local\Programs\Python\Python37\lib\site-packages\torch\nn\modules\module.py", line 570, in _apply
    module._apply(fn)
  File "C:\Users\gganesan166268\AppData\Local\Programs\Python\Python37\lib\site-packages\torch\nn\modules\module.py", line 570, in _apply
    module._apply(fn)
  File "C:\Users\gganesan166268\AppData\Local\Programs\Python\Python37\lib\site-packages\torch\nn\modules\module.py", line 570, in _apply
    module._apply(fn)
  [Previous line repeated 1 more time]
  File "C:\Users\gganesan166268\AppData\Local\Programs\Python\Python37\lib\site-packages\torch\nn\modules\module.py", line 593, in _apply
    param_applied = fn(param)
  File "C:\Users\gganesan166268\AppData\Local\Programs\Python\Python37\lib\site-packages\torch\nn\modules\module.py", line 680, in <lambda>
    return self._apply(lambda t: t.cuda(device))
  File "C:\Users\gganesan166268\AppData\Local\Programs\Python\Python37\lib\site-packages\torch\cuda\__init__.py", line 208, in _lazy_init
    raise AssertionError("Torch not compiled with CUDA enabled")
AssertionError: Torch not compiled with CUDA enabled
```
Is it that I have not installed the GPU version of cuda?

**Reply by fg-mindee** [reply]
It most likely means that you did not install the GPU version of PyTorch. Usually, it's because of some NVidia driver issue when you installed the package, and at that time it doesn't detect CUDA and picks the CPU-only build!

**Reply by fg-mindee** [reply]
did you manage to fix this @gganes3 ?

**Reply by gganes3** [reply]
No. I am still trying @fg-mindee. By the way, can you just let me know how I can retrain my detection model to size(256,1024). Also, what is the data that you use for training the detection model?

**Reply by fg-mindee** [reply]
First things first, could you try to investigate your CUDA situation? About non-square size for text detection and the data, it's a completely different topic so I suggest to discuss this elsewhere :sweat_smile:
